lib/perl/Genome/Config.pm

```diff
@@ -0,0 +1,90 @@
+package Genome::Config;
+
+use strict;
+use warnings;
+
+use Genome::Carp qw(croakf);
+use Genome::ConfigSpec qw();
+use Path::Class qw();
+use YAML::Syck qw();
+
+sub get {
+    my $key = shift;
+    my $spec = spec($key);
+
+    my $value = _lookup_value($spec);
+    if (!$spec->has_default_value && !defined($value)) {
+        croakf('required config key missing: %s', $key);
+    }
+
+    if (!defined($value)) {
+        return $spec->default_value;
+    }
+
+    for my $v (@{$spec->validators}) {
+        if (my $error = $v->($value, $spec)) {
```

Comment:
It seems odd that a call to validator's validate method (well, coderef) returns a truthy value on failure and a falsey value on success. I do like how you're using it to sort of auto-generate error messages though.

I wonder if part of the interface of a validator should be an `error_message` class method or something? I also wonder if its worth running all validators rather than dying after the first failure? That way you don't fix one problem only to stumble into the next one when you try again. Obviously this is just pseudocode but maybe something like: 

````perl
my @errors = map { $_->error_message($value) } 
        grep { ! $_->is_valid($value) }
        $spec->validators;

croakf(join(@error_messages, "\n")) if @error_messages;
````

Unrelated, but writing that made me wish perl had fold/reduce in core instead of an external module.